core/src/test/java/ai/timefold/solver/core/impl/score/stream/collector/connected_ranges/ConnectedRangeSolvingTest.java

```diff
@@ -0,0 +1,137 @@
+        public Integer getStart() {
+            return start;
+        }
+
+        public void setStart(Integer start) {
+            this.start = start;
+        }
+
+        public Integer getEnd() {
+            return start == null ? null : start + 10;
+        }
+    }
+
+    @PlanningSolution
+    public static class Planner {
+        @PlanningScore
+        private HardMediumSoftScore score;
+
+        @ProblemFactCollectionProperty
+        private final List<Equipment> equipments = new ArrayList<>();
+
+        @PlanningEntityCollectionProperty
+        private final List<Job> jobs = new ArrayList<>();
+
+        @ValueRangeProvider
+        public CountableValueRange<Integer> getStartOffsetRange() {
+            return ValueRangeFactory.createIntValueRange(0, 100);
+        }
+
+        public Planner() {
+        }
+
+        public Planner(List<Equipment> equipments, List<Job> jobs) {
+            this.equipments.addAll(equipments);
+            this.jobs.addAll(jobs);
+        }
+    }
+
+    public static class MyConstraintProvider implements ConstraintProvider {
+        public MyConstraintProvider() {
+        }
+
+        @Override
+        public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+            return new Constraint[] { doNotOverAssignEquipment(constraintFactory) };
+        }
+
+        public Constraint doNotOverAssignEquipment(ConstraintFactory constraintFactory) {
+            return constraintFactory.forEach(Equipment.class)
+                    .join(Job.class, Joiners.equal(Equipment::id, Job::getRequiredEquipmentId))
+                    .groupBy((equipment, job) -> equipment, ConstraintCollectors.toConnectedRanges((equipment, job) -> job,
+                            Job::getStart,
+                            Job::getEnd,
+                            (a, b) -> b - a))
+                    .flattenLast(ConnectedRangeChain::getConnectedRanges)
+                    .filter((equipment, connectedRange) -> connectedRange.getMaximumOverlap() > equipment.capacity())
+                    .penalize(HardMediumSoftScore.ONE_HARD)
+                    .asConstraint("Concurrent equipment usage over capacity");
+        }
+    }
+
+    @Test
+    public void solveConnectedRanges() {
+        var e1 = new Equipment(1, 1);
+        var j1 = new Job(1, e1.id());
+        var j2 = new Job(2, e1.id());
+        var problem = new Planner(List.of(e1), List.of(j1, j2));
+
+        var config = new SolverConfig()
+                .withSolutionClass(Planner.class)
+                .withEntityClasses(Job.class)
+                .withConstraintProviderClass(MyConstraintProvider.class)
+                .withEnvironmentMode(EnvironmentMode.FULL_ASSERT)
+                .withTerminationConfig(
+                        new TerminationConfig()
+                                .withScoreCalculationCountLimit(1_000L));
+        var solver = SolverFactory.create(config).buildSolver();
+        solver.solve(problem);
```

Comment:
Please write a test which tests the collector directly.

I think we have been overdoing it with these integration tests lately. There are several reasons why tests that run the solver should be the exception, rather than the norm:

- They take a relatively long time and if overused, our test suite will take impractically long to run.
- They do not actually test anything. Yes, the test otherwise would've failed without this fix, but that's not a guarantee of anything into the future. If the solver changes how it iterates through things, because for example the RNG impl in the JDK changed, the test may no longer be testing the same conditions and therefore no longer catch the issue.

These kinds of tests should only be used in cases where it is not possible or practical to reproduce a particular behavior on a lower level. There are cases like that, but this is not one of them - we know how to test collectors directly, and we should take the time to reproduce the actual problematic behavior.